### PR TITLE
Backwards incompatible changes to the API

### PR DESCRIPTION
**Pull request by cheddar**
Make a bunch of changes to the API.  The primary goal of the refactoring is to enable composition of HttpClient objects to effect both the RequestBuilder objects and the actual request processing.

There were two functionalities built into HttpClient that fit this mold, one was the credential handling and the other was the "enforce SSL" flag.  I have moved both of them to be wrappers on HttpClient objects.

This is important for me because it will enable me to decorate HttpClient objects with deployment-specific wrappers in Druid that shouldn't muddy up the Druid code proper.

On the actual API changes, I'm not necessarily married to them.  It's kinda weird that HttpClient is used to create the RequestBuilder and then subsequently must be passed in to the RequestBuilder to actually run the request.  I opted for that instead of separating the two interfaces because I figured that users of the client would prefer to just pass around one client instead of passing around a RequestBuilderFactory *and* an HttpClient.

The other option I can think of would be to make the Request object itself mutable and just eliminate the builder.

### CODE REVIEW

**Comment by gianm** [comment]
The way the decorators are implemented ends up looking strange- sometimes overriding "requestBuilder" and sometimes "go". And the Credentialed one in particular also seems easy to use wrong, since its "go" method will accept a request from RequestBuilder.post() that won't actually have the right credentials stuffed into it.

Would it work to make the HttpClient have just one method?

```java
   public <Intermediate, Final> ListenableFuture<Final> go(
       Request request,
       HttpResponseHandler<IntermediateType, Final> responseHandler
  )
```

where Request is an immutable thing that is not tied to any particular client or handler, and has some convenient methods for creating copies-with-modifications. The ssl-enforcing client would look at the current Request, validate it, and pass it down. The credentials-adding client would pass down a modified one with new headers. The netty client would actually make the request.

I think this is similar to your Request-being-mutable idea except it makes it easier for people to re-use Requests if they want (without worrying about it getting modified by the HttpClient)

**Comment by cheddar** [comment]
Yeah, that's basically the "Request is mutable" idea.  I can definitely do that, I just kinda fear that it could lead to a lot of copying of Multimaps for header updates.  Given that we are about to do a network call, maybe it's not worth worrying about, but it seems like a lot of extra garbage...

**Comment by gianm** [comment]
My assumption is that for the stuff we usually use http-client for, the copying of the header map won't be terribly noticeable, because we are usually dealing with heavy calls that have pretty large requests and responses.

Although I'm also fine documenting the interface with something like "your Request is potentially going to be modified" and making it mutable.

**Comment by cheddar** [comment]
@gianm 

I have redone it with a mutable request object and added a `copy()` method to the `Request` class.  This will allow the call site to determine if it prefers to have the cost of copying or the cost of mutation.

File: src/main/java/com/metamx/http/client/CredentialedHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package com.metamx.http.client;
```
Comment (gianm) [inline src/main/java/com/metamx/http/client/CredentialedHttpClient.java]: License

File: src/main/java/com/metamx/http/client/EnforceSslHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package com.metamx.http.client;
```
Comment (gianm) [inline src/main/java/com/metamx/http/client/EnforceSslHttpClient.java]: License

File: src/main/java/com/metamx/http/client/Request.java [hunk 3]
```diff
@@ -1,35 +1,53 @@
 package com.metamx.http.client;
```
Comment (gianm) [inline src/main/java/com/metamx/http/client/Request.java]: License

File: src/main/java/com/metamx/http/client/Request.java [hunk 4]
```diff
@@ -57,8 +75,103 @@ public ChannelBuffer getContent()
...
+
+  public Request setContent(String contentType, byte[] bytes, int offset, int length)
+  {
+    return setContent(contentType, factory.getBuffer(bytes, offset, length));
+  }
+
+  public Request setContent(String contentType, ChannelBuffer content)
+  {
+    if (contentType != null) {
+      addHeader(HttpHeaders.Names.CONTENT_TYPE, contentType);
```
Comment (gianm) [inline src/main/java/com/metamx/http/client/Request.java]: Should be setHeader, there can only be one content-type.

File: src/main/java/com/metamx/http/client/Request.java [hunk 5]
```diff
@@ -57,8 +75,103 @@ public ChannelBuffer getContent()
     return content;
   }
 
-  public HttpResponseHandler<Intermediate, Final> getHandler()
+  public Request copy() {
+    Request retVal = new Request(method, url);
+    retVal.headers.putAll(this.headers);
```
Comment (gianm) [inline src/main/java/com/metamx/http/client/Request.java]: This doesn't do a deep copy of the Objects in headers so they are still shared references. Can we change Object to String to fix that? Netty is mostly just gonna toString them anyway. (It seems to have exceptions for Date and Calendar)

**Comment by cheddar** [comment]
@gianm Addressed your concerns, please check again.

**Comment by gianm** [comment]
Looks good to me!

**Pull request merged**